File: diKoding/XamlMvvmBasic/XamlMvvmBasicVb/Person.vb
Imports System.ComponentModel
Imports System.Runtime.CompilerServices

Public Class Person
    Implements INotifyPropertyChanged

    Private _firstName As [String]

    Private _lastName As [String]

    Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

    Public Property FirstName As [String]
        Get
            Return _firstName
        End Get

        Set(value As [String])
            _firstName = value
            OnPropertyChanged()
            OnPropertyChanged("Name")
        End Set
    End Property

    Public Property LastName As [String]
        Get
            Return _lastName
        End Get

        Set(value As [String])
            _lastName = value
            OnPropertyChanged()
            OnPropertyChanged("Name")
        End Set
    End Property

    Public ReadOnly Property Name As [String]
        Get
            Return [String].Format("{0} {1}", Me._firstName, Me._lastName)
        End Get
    End Property

    Protected Overridable Sub OnPropertyChanged(<CallerMemberName> Optional propertyName As [String] = Nothing)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
    End Sub
End Class

File: diKoding/XamlMvvmBasic/XamlMvvmBasicCs/Person.cs
namespace XamlMvvmBasic
{
    using System;
    using System.ComponentModel;
    using System.Runtime.CompilerServices;
    
    public class Person
        : INotifyPropertyChanged
    {
        private String firstName;

        private String lastName;

        public event PropertyChangedEventHandler PropertyChanged;

        public String FirstName
        {
            get
            {
                return this.firstName;
            }

            set
            {
                this.firstName = value;
                this.OnPropertyChanged();
                this.OnPropertyChanged(@"Name");
            }
        }

        public String LastName
        {
            get
            {
                return this.lastName;
            }

            set
            {
                this.lastName = value;
                this.OnPropertyChanged();
                this.OnPropertyChanged(@"Name");
            }
        }

        public String Name
        {
            get
            {
                return 
                    String.Format(
                        @"{0} {1}", 
                        this.firstName, 
                        this.lastName);
            }
        }
        
        protected virtual void OnPropertyChanged(
            [CallerMemberName] String propertyName = null)
        {
            var handler = this.PropertyChanged;

            if (handler != null)
            {
                handler(this, new PropertyChangedEventArgs(propertyName));
            }
        }
    }
}